feat: Allow expressions as `group_by` keys

FBruzzesi opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues

- Closes #1385
- https://github.com/narwhals-dev/narwhals/issues/2225#issuecomment-2746592849
- https://github.com/narwhals-dev/narwhals/issues/2225#issuecomment-2747281275

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below

@MarcoGorelli there is [_one_ test failing](https://github.com/narwhals-dev/narwhals/blob/b940b2cbadd3d35d0f8813cbb5b1830f43480f2f/tests/frame/select_test.py#L57-L58) - which is grouping by a pandas column with a boolean name (and similarly with integer name in scikit-lego)

@dangotbanned I am crying for help with the typing 🙈

Other comment will address in the code itself

FBruzzesi commented on narwhals/_polars/group_by.py:
```diff
@@ -19,23 +20,36 @@
-        self._keys = list(keys)
-        df = df.drop_nulls(keys) if drop_null_keys else df
-        self._grouped: NativeGroupBy = df._native_frame.group_by(keys)
+        if not drop_null_keys:
+            self._compliant_frame = compliant_frame
+            self._keys = [extract_native(arg) for arg in keys]
+        else:
+            compliant_frame = compliant_frame.with_columns(*keys)
+            self._keys = [extract_native(arg).meta.output_name() for arg in keys]
+            self._compliant_frame = compliant_frame.drop_nulls(self._keys)
+
+        self._grouped: NativeGroupBy = self._compliant_frame._native_frame.group_by(
+            self._keys
+        )
```
So polars ended up being the ugliest, since there is not group_by param to deal with dropping nulls, but we support this (introduced for plotly)

FBruzzesi commented on narwhals/dataframe.py:
```diff
@@ -2826,17 +2830,20 @@ def group_by(
-            raise NotImplementedError(msg)
+        flat_keys, kinds = self._flatten_and_extract(*keys)
+
+        if any(kind is ExprKind.FILTRATION for kind in kinds):
+            from narwhals.exceptions import ShapeError
+
+            msg = "series used as keys should have the same length as the DataFrame"
+            raise ShapeError(msg)
```
That's what polars raises (when collecting) - should we post-pone it somehow?

MarcoGorelli replied:
why post-pone? seems good to raise early here

FBruzzesi commented on tests/group_by_test.py:
```diff
@@ -413,12 +414,6 @@ def test_all_kind_of_aggs(
     assert_equal_data(result, expected)
 
 
-def test_group_by_expr(constructor: Constructor) -> None:
-    df = nw.from_native(constructor({"a": [1, 1, 3], "b": [4, 5, 6]}))
-    with pytest.raises(NotImplementedError, match=r"not \(yet\?\) supported"):
-        df.group_by(nw.col("a")).agg(nw.col("b").mean())  # type: ignore[arg-type]
```
🎻

dangotbanned commented:
https://github.com/narwhals-dev/narwhals/actions/runs/14204180755/job/39797880528?pr=2325


> @dangotbanned I am crying for help with the typing 🙈


@FBruzzesi 😂 I'll try taking a look soon

dangotbanned commented:
https://github.com/narwhals-dev/narwhals/pull/2325#issuecomment-2770506951

@FBruzzesi for `polars` (**only**) just add an `Incomplete` in places where you need to use `CompliantExpr`.

https://github.com/narwhals-dev/narwhals/blob/4d2b9d57f70e52b7c78ca3f41e228be2ceb96cfa/narwhals/_polars/namespace.py#L54-L55

`PolarsSeries` and `PolarsExpr` are *almost entirely* defined using `__getattr__` - which isn't enough to satisfy typing.

They essentially need something like this, but I've been putting off doing it since they have waaaaaay more methods 😔 

https://github.com/narwhals-dev/narwhals/blob/4d2b9d57f70e52b7c78ca3f41e228be2ceb96cfa/narwhals/_polars/dataframe.py#L59-L85

dangotbanned commented on narwhals/_compliant/series.py:
```diff
@@ -300,7 +300,7 @@ def _with_native(self, series: Any, *, preserve_broadcast: bool = False) -> Self
 
         In cases when operations are known to not affect whether a result should
-        be broadcast, we can pass `preverse_broadcast=True`.
+        be broadcast, we can pass `preserve_broadcast=True`.
```
> preverse

😄

dangotbanned commented on narwhals/_arrow/group_by.py:
```diff
@@ -41,18 +42,26 @@ class ArrowGroupBy(EagerGroupBy["ArrowDataFrame", "ArrowExpr"]):
-        else:
-            self._compliant_frame = compliant_frame
-        self._keys: list[str] = list(keys)
+        compliant_frame = compliant_frame.with_columns(*keys)
+
+        self._keys: list[str] = list(
+            chain.from_iterable(
+                evaluate_output_names_and_aliases(
+                    expr=key, df=compliant_frame, exclude=[]
+                )[1]
+                for key in keys
+            )
+        )
+        self._compliant_frame = (
+            compliant_frame.drop_nulls(self._keys) if drop_null_keys else compliant_frame
+        )
```
Is it alright if I take a look at this tomorrow?

At a glance, I don't think it needs to be this complex

FBruzzesi replied:
Of course, there is no rush!

dangotbanned commented on narwhals/_dask/group_by.py:
```diff
@@ -72,12 +73,24 @@ class DaskLazyGroupBy(DepthTrackingGroupBy["DaskLazyFrame", "DaskExpr", Aggregat
     ) -> None:
-        self._compliant_frame = df
-        self._keys: list[str] = list(keys)
+        self._compliant_frame = compliant_frame.with_columns(*keys)
+        self._keys: list[str] = list(
+            chain.from_iterable(
+                evaluate_output_names_and_aliases(
+                    expr=key, df=compliant_frame, exclude=[]
+                )[1]
+                for key in keys
+            )
+        )
```
This looks identical to (https://github.com/narwhals-dev/narwhals/pull/2325/files#r2023601119) - which makes me think that part could be defined in `DepthTrackingGroupBy`

dangotbanned commented on narwhals/_duckdb/group_by.py:
```diff
@@ -17,14 +18,27 @@
-        self._keys = list(keys)
+        compliant_frame = compliant_frame.with_columns(*keys)
+
+        self._keys: list[str] = list(
+            chain.from_iterable(
+                evaluate_output_names_and_aliases(
+                    expr=key, df=compliant_frame, exclude=[]
+                )[1]
+                for key in keys
+            )
+        )
```
Third? https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2023601119

dangotbanned commented on narwhals/_pandas_like/group_by.py:
```diff
@@ -40,20 +41,30 @@ class PandasLikeGroupBy(EagerGroupBy["PandasLikeDataFrame", "PandasLikeExpr"]):
-        self._compliant_frame = df
-        self._keys: list[str] = list(keys)
+        self._compliant_frame = compliant_frame.with_columns(*keys)
+        self._keys: list[str] = list(
+            chain.from_iterable(
+                evaluate_output_names_and_aliases(
+                    expr=key, df=compliant_frame, exclude=[]
+                )[1]
+                for key in keys
+            )
+        )
```
Fourth? https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2023601119

dangotbanned commented:
https://github.com/narwhals-dev/narwhals/pull/2325#event-17081967098

😨

FBruzzesi commented on narwhals/_spark_like/group_by.py:
```diff
@@ -16,14 +18,26 @@
-        self._compliant_frame = df.drop_nulls(subset=None) if drop_null_keys else df
-        self._keys = list(keys)
+        compliant_frame = compliant_frame.with_columns(*keys)
+        self._keys: list[str] = list(
+            chain.from_iterable(
+                evaluate_output_names_and_aliases(
+                    expr=key, df=compliant_frame, exclude=[]
+                )[1]
+                for key in keys
+            )
+        )
```
Ok I definitely need to catch up with all the new internals to avoid repetition 😂

dangotbanned replied:
Haha no worries, maybe the changes here will be helpful

- #2252

I have a feeling I did an overly-optimized version of this in `LazyGroupBy`

FBruzzesi commented on tests/frame/select_test.py:
```diff
@@ -59,7 +59,7 @@ def test_select_boolean_cols(request: pytest.FixtureRequest) -> None:
         request.applymarker(pytest.mark.xfail)
     df = nw.from_native(pd.DataFrame({True: [1, 2], False: [3, 4]}), eager_only=True)
-    result = df.group_by(True).agg(nw.col(False).max())  # type: ignore[arg-type]# noqa: FBT003
+    result = df.group_by(nw.col(True)).agg(nw.col(False).max())  # type: ignore[arg-type]# noqa: FBT003
```
@MarcoGorelli I think this is a fair "request". If we are going to support columns with a boolean value as a name, I think it would be ok to ask to wrap it in a `nw.col(True)` to disambiguate with `nw.lit(True)`

FBruzzesi replied:
On a second thought, this would actually be a breaking change that might require to update _all_/_most_ downstream libraries.

Just because plotly doesn't have a test for that, it doesn't mean that users don't pass a numeric named column 🙈

dangotbanned replied:
@FBruzzesi I agree with your first comment (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2024085160)

This isn't documented as something we allow - so personally I wouldn't expect to be able to rely on it

https://narwhals-dev.github.io/narwhals/api-reference/dataframe/#narwhals.dataframe.DataFrame.group_by

dangotbanned replied:
What is the use-case for having a column *valued* as `True`?

Even writing that seemed fishy 😄

FBruzzesi replied:
> What is the use-case for having a column _valued_ as `True`?

Trust me when I say that we all want to live in a world where that's not allowed, but for pandas is.

The example in scikit-lego might be a bit more realistic:

```py
import numpy as np
import pandas as pd

arr = np.array([[1,2], [3,4], [5,6]])
type(pd.DataFrame(arr).columns[0])
int
```

In the scikit-lego case we can control that (since we are creating the dataframe ourselves), but for plotly we can't. It's dataframe as input, column name as input

dangotbanned replied:
😔

dangotbanned replied:
Hang on, doesn't `validate_column_names` handle this for us?

FBruzzesi replied:
`validate_column_names` checks for:

- uniqueness 
- hashability 

of pandas column names. It does not have an effect on the column name types

dangotbanned replied:
Ah fair, thanks @FBruzzesi

FBruzzesi replied:
I know it's not a _great_ solution, but it's one that makes this backward compatible - Just pushed the following path in `_extract_compliant` methods:

```py
if self.implementation.is_pandas_like() or self.implementation.is_dask():
    if isinstance(arg, (int, bool)):
        return plx.col(arg)  # type: ignore[arg-type]
    if isinstance(arg, Sequence) and all(isinstance(x, (int, bool)) for x in arg):
        return plx.col(*arg)
```

I would propose to have it _only_ for v1, and v2 ask for explicitly pass `nw.col(...)` to control those cases. I understand that's not the most natural think in a `group_by` though.

Edit: I now see the test failing, since in the select context we where banning that already. I will add the special path in the group by only

dangotbanned commented:
- https://github.com/narwhals-dev/narwhals/pull/2325#issuecomment-2770506951
- https://github.com/narwhals-dev/narwhals/pull/2325#issuecomment-2770524042

@FBruzzesi these are *finally* done now in #2328 😅 

If that PR merges *before* this one - I can take another stab at the `polars` typing.
Should be a lot simpler

FBruzzesi commented on narwhals/dataframe.py:
```diff
@@ -1546,17 +1546,45 @@ def group_by(
-                "Hint: instead of `df.group_by(nw.col('a'))`, use `df.group_by('a')`."
-            )
-            raise NotImplementedError(msg)
+        if (
+            self.implementation.is_pandas_like() or self.implementation.is_dask()
+        ):  # pragma: no cover
+            from narwhals import col
+
+            def _maybe_parse_key_to_expr(
+                key: IntoExpr  # noqa: FBT001
+                | Iterable[IntoExpr]
+                | int
+                | Iterable[int]
+                | bool
+                | Iterable[bool],
+            ) -> IntoExpr:
+                if isinstance(key, (int, bool)):
+                    return col(key)  # type: ignore[arg-type]
+                elif isinstance(key, Iterable) and all(
+                    isinstance(k, (int, bool)) for k in key
+                ):
+                    return col(*key)
+                else:
+                    return key  # type: ignore[return-value]
+
+            keys = tuple(_maybe_parse_key_to_expr(key) for key in keys)
```
Happy to move this beauty in the v1 path only - LMK your thoughts

FBruzzesi replied:
Also, I am not a big fan of this function, I can think of many many edge cases

dangotbanned replied:
You can always just define it as `utils._maybe_parse_key_to_expr`?

dangotbanned replied:
Followed up in (https://github.com/narwhals-dev/narwhals/pull/2325/commits/3f93217bc3aa25be95ddcc93a3c8b2ef3baa7fe3)

Meant to write `_expression_parsing` 😄 

I've also removed the `# pragma: no cover`, as it doesn't make sense to me that we'd have code which both:
- Supports a case that isn't part of the API
- Doesn't have any tests

I would pick **one** of those personally 🙂

dangotbanned commented on narwhals/dataframe.py:
```diff
@@ -2826,17 +2854,45 @@ def group_by(
-                "Hint: instead of `df.group_by(nw.col('a'))`, use `df.group_by('a')`."
-            )
-            raise NotImplementedError(msg)
+        # Handle special case for pandas-like and dask, which allow column names to be
+        # non-string values
+        if (
+            self.implementation.is_pandas_like() or self.implementation.is_dask()
+        ):  # pragma: no cover
+            from narwhals import col
+
+            def _maybe_parse_key_to_expr(
+                key: IntoExpr  # noqa: FBT001
+                | Iterable[IntoExpr]
+                | int
+                | Iterable[int]
+                | bool
+                | Iterable[bool],
+            ) -> IntoExpr:
+                if isinstance(key, (int, bool)):
+                    return col(key)  # type: ignore[arg-type]
+                if isinstance(key, Iterable) and all(
+                    isinstance(k, (int, bool)) for k in key
+                ):
+                    return col(*key)  # pyright: ignore[reportArgumentType]
+                return key  # type: ignore[return-value]
+
+            keys = tuple(_maybe_parse_key_to_expr(key) for key in keys)
+
+        flat_keys, kinds = self._flatten_and_extract(*keys)
+
+        if any(kind is ExprKind.FILTRATION for kind in kinds):
+            from narwhals.exceptions import ShapeError
+
+            msg = "series used as keys should have the same length as the DataFrame"
+            raise ShapeError(msg)
```
Does any of this get *simpler* if we don't add support for `Series` as keys?

Would supporting `Series` *only* as keys for `DataFrame.group_by` and not `LazyFrame.group_by` be an option?

FBruzzesi replied:
Regarding the parsing part, I guess you already solved it 🚀🙌🏼

dangotbanned replied:
> Regarding the parsing part, I guess you already solved it 🚀🙌🏼

Oh nice one 😄

FBruzzesi commented on narwhals/_expression_parsing.py:
```diff
@@ -424,3 +425,30 @@ def apply_n_ary_operation(
+            yield col(arg)  # type: ignore[arg-type]
+        elif isinstance(arg, Iterable):
+            yield from _iter_group_by_keys_into_exprs(arg)
+        else:  # pragma: no cover
+            # pandas can have also tuple as column names
+            # TODO(FBruzzesi): What should we do in such case? In the current implementation
+            # that would enter the block above, and most likely end up in a `ColumnNotFoundError`
+            msg = f"Expected argument that can be converted into an expression, got: {type(arg)}. Please report a bug."
+            raise TypeError(msg)
```
Here is an example in practice:

```py
import pandas as pd

df_native = pd.DataFrame(
    {
        (1, "a"): [1, 2, 1],
        "b": [4.2, 3.14, 9.81]
    }
)

df = nw.from_native(df_native, eager_only=True)
df.group_by(pd.MultiIndex.from_tuples([(1, "a")]))
# and df.group_by((1, "a")) as well
```

> narwhals.exceptions.ColumnNotFoundError: The following columns were not found: [1]
> E           
> E           Hint: Did you mean one of these columns: [(1, 'a'), 'b', 'c']?

as:

```py
isinstance(pd.MultiIndex.from_tuples([(1, "a")]), Iterable)
True
```

dangotbanned replied:
@FBruzzesi passing a `tuple` as a column name would break quite a lot more things than just this.

Any usage of `utils.flatten` is entirely incompatible with that case.

*If* however a `tuple` managed to reach `_iter_group_by_keys_into_exprs` it would *either*`:


Be iterated over as `args`:
https://github.com/narwhals-dev/narwhals/blob/8f6632c5496cb1f3a9509fc40d53e0c06bf61ec0/narwhals/_expression_parsing.py#L435

Or trigger a recursive call here:

https://github.com/narwhals-dev/narwhals/blob/8f6632c5496cb1f3a9509fc40d53e0c06bf61ec0/narwhals/_expression_parsing.py#L440-L411

FBruzzesi replied:
My concern is mostly backward compatibility.

I would lean toward simplifying this completely by:
- do not support expr in V1
- allow expr in the main namespace, but don't allow ambiguous values such as booleans. It would be enough to wrap booleans, integers, tuples, etc.. with a `nw.col(...)`

dangotbanned replied:
@FBruzzesi 

Yeah that sounds good to me, so in `MAIN` we just support what is currently annotated?

https://github.com/narwhals-dev/narwhals/blob/8f6632c5496cb1f3a9509fc40d53e0c06bf61ec0/narwhals/dataframe.py#L1501-L1503

dangotbanned replied:
@FBruzzesi, @MarcoGorelli 
is there a possibility of dropping support for non-`str` columns **everywhere** in `v2`?

That seems like a more consistent rule than something that applies *only* in `df.group_by(...)`.

Totally understand if not - just thought I'd float the idea 😅

FBruzzesi commented:
Thanks @dangotbanned to push this forward. I think there is one remaining open point (see [comment](https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2030210542), after which this should be ready for review!

The failing test in nightly is unrelated

dangotbanned commented on narwhals/_expression_parsing.py:
```diff
@@ -439,8 +439,12 @@ def _iter_group_by_keys_into_exprs(args: Iterable[Any], /) -> Iterator[IntoExpr
-            yield arg
+        else:  # pragma: no cover
+            # pandas can have also tuple as column names
+            # TODO(FBruzzesi): What should we do in such case? In the current implementation
+            # that would enter the block above, and most likely end up in a `ColumnNotFoundError`
```
I think its fine for this to result in whatever the normal error is.

IIUC, what you wanted was to add some *exceptional* cases to skip the regular column parsing rules?

If that's the case - everything else should go through as normal IMO

FBruzzesi replied:
Resolving conversation as moved forward with the distinction between Main and V1 namespaces

FBruzzesi commented on narwhals/stable/v1/__init__.py:
```diff
@@ -250,6 +253,27 @@ def _l1_norm(self: Self) -> Self:
+                "\nUse main namespace class instead: `narwhals.DataFrame.group_by`"
+            )
+            raise ComputeError(msg)
+        return super().group_by(*[col(k) for k in keys], drop_null_keys=drop_null_keys)
```
Notice that if we were to do:

```py
return super().group_by(col(*keys), drop_null_keys=drop_null_keys)
```
which in principle is more pythonic/polars-like, then polars `extract_native(arg).meta.output_name()` itself would raise an exception

FBruzzesi commented:
@dangotbanned @MarcoGorelli let me summarize the PR state after so many commits together with Dan.

- Main namespace will allow any `IntoExpr` object as part of `group_by` keys
  - If a key is not parse-able into an expression, then and exception will be raised.
  - One case of this is what would be allowed generally in pandas, such as integers and booleans, unless wrapped into `nw.col(...)`
- V1 is maintained stable - therefore expressions _cannot_ be passed as keys, but integers and booleans can as of now.
- Large diff is mostly due to test suite changes

MarcoGorelli reviewed (commented):
thanks for exploring this!

this it out, i'm not sure it's right:
```python
In [20]: df
Out[20]: 
pyarrow.Table
a: int64
b: int64
----
a: [[1,1,2]]
b: [[4,5,6]]

In [21]: pl.from_arrow(df).group_by(pl.col('a').alias('b')).agg(pl.col('b').mean().alias('c'))
Out[21]: 
shape: (2, 2)
┌─────┬─────┐
│ b   ┆ c   │
│ --- ┆ --- │
│ i64 ┆ f64 │
╞═════╪═════╡
│ 2   ┆ 6.0 │
│ 1   ┆ 4.5 │
└─────┴─────┘

In [22]: nw.from_native(df).group_by(nw.col('a').alias('b')).agg(nw.col('b').mean().alias('c'))
Out[22]: 
┌──────────────────┐
|Narwhals DataFrame|
|------------------|
|  pyarrow.Table   |
|  b: int64        |
|  c: double       |
|  ----            |
|  b: [[1,2]]      |
|  c: [[1,2]]      |
└──────────────────┘
````

FBruzzesi commented:
Moving back to draft since there are a few cases I need to solve:

- [ ] output order
- [ ] sqlframe failing because of schema

@MarcoGorelli as from today's community call, I should have addressed (in the simplest way possible) v1 vs main namespaces discrepancy

FBruzzesi commented on narwhals/_compliant/group_by.py:
```diff
@@ -67,14 +70,43 @@ def compliant(self) -> CompliantFrameT_co:
     def agg(self, *exprs: CompliantExprT_contra) -> CompliantFrameT_co: ...
 
+    @staticmethod
+    def _magic_parsing(
```
Whops! that was not supposed to be pushed 😅

dangotbanned replied:
Naming aside ...

I'm having a hard time understanding how *this much* complexity is needed 🤔

FBruzzesi replied:
Hey Dan! I am fairly cooked for the day so I hope this does not sound confusing.
Given Marco's example in the [comment](https://github.com/narwhals-dev/narwhals/pull/2325#pullrequestreview-2757684799), the _strategy_ to replicate such behaviour is other dataframes is:

- create columns with the expressions in the keys, however:
  - make sure that the names do not overlap (hence overwrite) any existing column
  - remove the original key column _source_, as otherwise something like `df.group_by(nw.col("x")).agg(nw.all().sum())` would end up considering column `"x"`, which should only be in the keys
- track output name in order to remap the temporary column names to the actual expected output name

Happy to brainstorm to figure out a better solution for the case Marco brought up

FBruzzesi replied:
I had the idea of introducing a fast path with a condition which looks like the following (via function name or depth)

```py
# fastpath for the case in which all keys are `nw.col(key)`
if all(getattr(k, "_function_name", None) == "col" for k in keys):
    output_names = compliant_frame._evaluate_aliases(*keys)
    return compliant_frame, output_names, output_names
```
but then realized that aliases are problematic since depth nor function name are tracked.

I will think about this after solving the remaining issues

FBruzzesi commented on narwhals/_compliant/group_by.py:
```diff
@@ -67,14 +70,43 @@ def compliant(self) -> CompliantFrameT_co:
+        output_names = compliant_frame._evaluate_aliases(*keys)
+
+        safe_keys = [
+            key.name.suffix(suffix_token)
+            if key._metadata is not None
+            and key._metadata.expansion_kind is ExpansionKind.MULTI_NAMED
+            else key.alias(f"{new_name}{suffix_token}")
```
If it's multinamed, then it cannot be duplicate, therefore as temporary name we can just use the original one with a suffix. Otherwise, we alias with a suffix

dangotbanned commented on narwhals/_polars/group_by.py:
```diff
@@ -20,44 +20,56 @@
+            by = [extract_native(arg) for arg in keys]
+            self._compliant_frame = df
+        else:
+            by = [extract_native(arg).meta.output_name() for arg in keys]
```
```suggestion
            by = [arg.native.meta.output_name() for arg in keys]
```

dangotbanned commented on narwhals/_polars/group_by.py:
```diff
@@ -20,44 +20,56 @@
-        return from_native(self._grouped.agg(extract_native(arg) for arg in aggs))
+    def agg(self, *aggs: PolarsExpr) -> PolarsDataFrame:
+        return self.compliant._with_native(
+            self._grouped.agg(extract_native(arg) for arg in aggs)
```
```suggestion
            self._grouped.agg(arg.native for arg in aggs)
```

dangotbanned commented on narwhals/_polars/group_by.py:
```diff
@@ -20,44 +20,56 @@
-        return from_native(self._grouped.agg(extract_native(arg) for arg in aggs))
+        by: Sequence[pl.Expr | str]
+        if not drop_null_keys:
+            by = [extract_native(arg) for arg in keys]
```
```suggestion
            by = [arg.native for arg in keys]
```

dangotbanned commented on narwhals/_polars/group_by.py:
```diff
@@ -20,44 +20,56 @@
+            by = [extract_native(arg) for arg in keys]
+            self._compliant_frame = df
+        else:
+            by = [extract_native(arg).meta.output_name() for arg in keys]
```
```suggestion
            by = [arg.native.meta.output_name() for arg in keys]
```

dangotbanned commented on narwhals/_polars/group_by.py:
```diff
@@ -20,44 +20,56 @@
+
+    def agg(self, *aggs: PolarsExpr) -> PolarsLazyFrame:
+        return self.compliant._with_native(
+            self._grouped.agg(extract_native(arg) for arg in aggs)
```
```suggestion
            self._grouped.agg(arg.native for arg in aggs)
```

FBruzzesi commented on tests/group_by_test.py:
```diff
@@ -431,7 +425,9 @@ def test_pandas_group_by_index_and_column_overlap() -> None:
     key, result = next(iter(nw.from_native(df, eager_only=True).group_by("a")))
     assert key == (1,)
-    expected_native = pd.DataFrame({"a": [1, 1], "b": [4, 5]})
+    expected_native = pd.DataFrame(
+        {"a": [1, 1], "b": [4, 5]}, index=pd.Index([0, 1], name="a")
+    )
```
Arguably this is the expected behaviour since:

```py
import pandas as pd

data = {"a": [1, 1, 1, 2], "b": [4, 3, None, None]}
frame = pd.DataFrame(data, index=pd.Index([0,1,2,3], name="foo"))

for key, grp in frame.groupby("a"):
    print(key, grp.index.name)

# 1 foo
# 2 foo

frame.groupby("a", as_index=False)["b"].sum().index.name is None
# True
```

FBruzzesi commented:
Failing tests:

- [ ] python 3.13, arguably an issue upstream with polars streaming engine. I will try to reproduce and report it upstream
- [ ] polars pre 1.0 does no support expr in keys?
- [x] shiny - unrelated
- [ ] validoopsie - requires investigation
- [ ] mypy 🥲

FBruzzesi commented on narwhals/_polars/dataframe.py:
```diff
@@ -373,10 +372,15 @@ def to_dict(
         from narwhals._polars.group_by import PolarsGroupBy
 
-        return PolarsGroupBy(self, keys, drop_null_keys=drop_null_keys)
+        return cast(
+            "CompliantGroupBy[PolarsDataFrame, Any]",
+            PolarsGroupBy(self, keys, drop_null_keys=drop_null_keys),
+        )
```
@dangotbanned does this make sense? (similarly in the lazy case) 🙈

dangotbanned replied:
@FBruzzesi If you need to cast it, then there's an issue elsewhere

I haven't checked the PR, but if you've changed the definitions of one of the other protocols - then that could be the issue?

FBruzzesi replied:
Ohh I see! I had to add a few attributes I believe. I will check after dinner

dangotbanned replied:
Please don't merge this without resolving that issue 🙏

FBruzzesi commented on narwhals/_duckdb/dataframe.py:
```diff
@@ -242,18 +242,15 @@ def _with_native(self: Self, df: duckdb.DuckDBPyRelation) -> Self:
-            for name in df.columns
-        ]
-        return self._with_native(df.select(", ".join(selection)))
+        selection = (col(name).alias(mapping.get(name, name)) for name in df.columns)
+        return self._with_native(self.native.select(*selection))
```
The issue with existing implementation is for dashed names. If mapping is `{"foo": "foo-bar"}`, then the selection evaluates to `"foo as foo-bar"` and ends up failing!

MarcoGorelli replied:
ooh nice! daily reminder to use the python api more and handwritten sql less 😄

MarcoGorelli replied:
can we make a separate PR to fix the issue with the current implementation, so we can merge that immediately?

FBruzzesi replied:
Sure thing!

FBruzzesi replied:
There you go: #2382

dangotbanned commented on tests/group_by_test.py:
```diff
@@ -11,6 +11,9 @@
 
 import narwhals.stable.v1 as nw
+from narwhals._expression_parsing import ExprKind
+from narwhals._expression_parsing import infer_kind
```
I don't think it's a good idea to include the internals in the tests

You run the risk of just testing the current behaviour, rather than what is expected.

Do you think you could form the condition without this?

FBruzzesi replied:
Solved in https://github.com/narwhals-dev/narwhals/pull/2325/commits/5bf2179a310d1627fa53b07a377adfa761f64104

MarcoGorelli commented:
thanks @FBruzzesi !

another test case (sorry! 🙈 )

```python
import polars as pl

import pandas as pd
import narwhals as nw

data = {"a": [1,1,2], "b": [4,5,6]}
df = nw.from_native(pd.DataFrame(data)).lazy('duckdb')
print(df.group_by((nw.col('a')*2).alias('b')).agg(nw.all().sum().name.suffix('c')))
df = pl.DataFrame(data)
print(df.group_by((pl.col('a')*2).alias('b')).agg(pl.all().sum().name.suffix('c')))
```

```
┌──────────────────┐
|Narwhals LazyFrame|
|------------------|
|┌───────┬────────┐|
|│   b   │   bc   │|
|│ int64 │ int128 │|
|├───────┼────────┤|
|│     2 │      9 │|
|│     4 │      6 │|
|└───────┴────────┘|
└──────────────────┘
shape: (2, 2)
┌─────┬─────┐
│ b   ┆ ac  │
│ --- ┆ --- │
│ i64 ┆ i64 │
╞═════╪═════╡
│ 2   ┆ 2   │
│ 4   ┆ 2   │
└─────┴─────┘
```

MarcoGorelli commented:
Just to demistify the Polars case above:
```
df.group_by((pl.col('a')*2).alias('b')).agg(pl.all().sum().name.suffix('c'))
```
- `pl.all().sum()` is a multi-output unnamed expression
- it would ordinarily get expanded to `[pl.col('a').sum(), pl.col('b').sum()]`
- but here, we exclude the key(s), so it only gets expanded to `[pl.col('a').sum()]`
- So, the result contains 'b' (the key) and the sum of column 'a' per group, suffixed 'c'

dangotbanned commented:
@FBruzzesi I don't mean to throw *too many* spanners in the works here 😅 but ...

Have you considered that all *current* users of `(Data|Lazy)Frame.group_by` are passing `str` or `Sequence[str]`?

That information is lost (*or at least made less-accessible*) by immediately using `col` and then evaluating.

https://github.com/narwhals-dev/narwhals/blob/0c1d42709379fbade2992fdfaa78d3bc649cb69a/narwhals/dataframe.py#L1560-L1569


So existing code would incur a performance cost (*despite not using expressions*) by unconditionally going through this section:

https://github.com/narwhals-dev/narwhals/blob/0c1d42709379fbade2992fdfaa78d3bc649cb69a/narwhals/_compliant/group_by.py#L83-L115


> [!TIP]
> I think you should look at identifying fastpaths at the Narwhals-level

For example, if we know that `keys: tuple[str, ...]` - then we could call an **alternative constructor** for `GroupBy`:

```py
class GroupBy:
    @classmethod
    def from_exprs(cls, ...) -> GroupBy: ...
    @classmethod
    def from_strings(cls, ...) -> GroupBy: ...
    @classmethod
    def from_i_dunno_man(cls, ...) -> GroupBy: ...
```

FBruzzesi commented:
> So existing code would incur a performance cost (_despite not using expressions_) by unconditionally going through this section:

Hey @dangotbanned, yes this is definitly in the back of my mind (see [comment](https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2040610045)).

All the tests are now passing, except: I think polars had a regression somewhere between v0.x and v1.x for not allowing multinamed expression as keys.

Hopefully sometime next week I will finish the PR 😅 Let me move it back to draft in the meanwhile

dangotbanned commented:
https://github.com/narwhals-dev/narwhals/pull/2325#issuecomment-2800113898

No worries @FBruzzesi, it looks like I 👀'd (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2040610045) already 😄 

Just wanted to check in, couldn't tell if the plan was to get this shipped for tomorrow.
I'll retreat back to the sidelines for now 😉

FBruzzesi commented:
> [#2325 (comment)](https://github.com/narwhals-dev/narwhals/pull/2325#issuecomment-2800113898)
> 
> No worries @FBruzzesi, it looks like I 👀'd ([#2325 (comment)](https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2040610045)) already 😄
> 
> Just wanted to check in, couldn't tell if the plan was to get this shipped for tomorrow. I'll retreat back to the sidelines for now 😉

Ideas are welcomed as I didn't put much thought into it.
The `from_<keys_type>` pattern is already a great idea that might work in practice. I will explore it further

FBruzzesi commented:
> ```python
> @classmethod
>     def from_exprs(cls, ...) -> GroupBy: ...
>     @classmethod
>     def from_strings(cls, ...) -> GroupBy: ...
> ```

@dangotbanned I have a quite simple solution in place - here is my reasoning: as `GroupBy` and `LazyGroupBy` are potentially user facing, I didn't want to add methods to those classes. Additionally it seemed a bit of an overkill since all I needed was <10 lines of code.

I am checking if all keys are strings, otherwise _cast_ everything to an expression and pass it along. Two drawbacks here:
- mypy is screaming, _a lot_. So maybe having different methods to instantiate the class would help here
- I am checking for all elements being strings twice - once to return early in `DataFrame.group_by`, once at the compliant level to assign values.

I am going to push it for now and iterate on that to address these two points if it generally looks good

dangotbanned commented:
> Tests are failing - it took a while to figure out the reason being: [`super().__init__` doesn't work in explicit protocol subclasses](https://github.com/python/typing/issues/572)


@FBruzzesi 
Aha now someone else has stumbled into the same issue I had 😄

See this for options (note that `from_*` is also good for this):
- https://github.com/narwhals-dev/narwhals/pull/2294#discussion_r2014534830

FBruzzesi commented:
> > Tests are failing - it took a while to figure out the reason being: [`super().__init__` doesn't work in explicit protocol subclasses](https://github.com/python/typing/issues/572)
> 
> @FBruzzesi Aha now someone else has stumbled into the same issue I had 😄
> 
> See this for options (note that `from_*` is also good for this):
> 
> * [chore: Add `EagerSeries*Namespace` protocols #2294 (comment)](https://github.com/narwhals-dev/narwhals/pull/2294#discussion_r2014534830)
> 
> I can try taking a look tomorrow?
> 
> I'm thinking something like adding a `_group_by` property to each of `CompliantDataFrame` and `CompliantLazyFrame`. Then you could add whatever constructors make sense, but avoid the issues with `__init__` on old pythons. That would also make the typing a bit easier, since you could use `CompliantGroupBy[Self, CompliantExprT_contra]` from the frame class, where that's already in scope

@dangotbanned sorry I mostly missed this and today I didn't have the time to look at anything Narwhals related!

I ended up rolling back to a semi-private method and "split" it into two staticmethods `_parse_string_keys` and `_parse_expr_keys`.

I am tempted to marking it ready to review since the performance concerns should be addressed, and the only remaining issue is the type annotation 🙈

MarcoGorelli commented on narwhals/_spark_like/dataframe.py:
```diff
@@ -187,7 +187,7 @@ def _iter_columns(self) -> Iterator[Column]:
 
     @property
     def columns(self: Self) -> list[str]:
-        return list(self.schema)
+        return list(self.native.columns)
```
`self.schema` gets cached, can we revert this?

that, and perhaps add a comment explaining why we use `self.schema` in case anyone changes this in the future

FBruzzesi replied:
There is an issue with both SQLFrame when using `self.schema`: in `SparkLikeLazyGroupBy.agg`, when getting to the last line, `.rename` fails due to:

> NotImplementedError: This engine does not support schema inference likely since it does not have an active connection.

which I didn't really bother to investigate deeper considering all the remaining issues and edge case of the PR.

Maybe we could even do:

```py
return list(self.schema) if self._cached_schema else list(self.native.columns)
```

Aside the failing case, why should we collect the schema when only for column names?

MarcoGorelli replied:
> Maybe we could even do:

hmmm yeah that looks nice, and then also do the same caching for `.columns`? we call `.columns` a few times in `join`, probably good to cache it. but, ok for a separate pr

MarcoGorelli commented on narwhals/dataframe.py:
```diff
@@ -1542,18 +1543,27 @@ def group_by(
+        if not all(kind is ExprKind.TRANSFORM for kind in kinds):
+            from narwhals.exceptions import ComputeError
+
+            msg = "Group by is not (yet) supported with keys that are not transformation expressions"
```
I don't think we will ever support this, right? can we remove the `'(yet)'`?

MarcoGorelli commented on tests/expr_and_series/reduction_test.py:
```diff
@@ -26,7 +26,6 @@
             {"a": [1, 2, 3], "min": [4, 4, 4]},
         ),
     ],
-    ids=range(5),
```
why remove these?

FBruzzesi replied:
It was unused in this test

MarcoGorelli replied:
they are pytest ids https://docs.pytest.org/en/stable/example/parametrize.html

though admittedly `range(5)` aren't very helpful ids 😄

FBruzzesi replied:
I know, most likely I introduced them😂  I spotted they were unused while I was introducing for a new test case (I did not recall the syntax, but I knew we had it in our test suite).

They where previously used to xfail certain tests via the id value identifier. I believe that once you added full support for aggregations in pyspark and duckdb the xfail-ures where removed and these ids unused!

MarcoGorelli commented on narwhals/_compliant/group_by.py:
```diff
@@ -67,12 +71,69 @@ def compliant(self) -> CompliantFrameT_co:
+            no overlap with any existing column name.
+        - Add these temporary columns to the compliant dataframe.
+        """
+        suffix_token = "_" * max(len(str(c)) for c in compliant_frame.columns)
```
can you add 1 to the max? `''` is a valid column name

dangotbanned commented on narwhals/_compliant/group_by.py:
```diff
@@ -67,12 +71,69 @@ def compliant(self) -> CompliantFrameT_co:
         drop_null_keys: bool,
     ) -> None: ...
 
+    def _parse_keys(
+        self,
+        compliant_frame: CompliantLazyFrameAny,
+        keys: Sequence[CompliantExprT_contra] | Sequence[str],
+    ) -> tuple[CompliantLazyFrameAny, list[str], list[str]]:
+        if all(isinstance(k, str) for k in keys):
+            keys = cast("list[str]", list(keys))
+
+            return self._parse_string_keys(compliant_frame, keys=keys)
+        else:
+            keys = cast("list[CompliantExprT_contra]", list(keys))
+            return self._parse_expr_keys(compliant_frame, keys=keys)
+
+    @staticmethod
+    def _parse_expr_keys(
+        compliant_frame: CompliantLazyFrameAny, keys: Sequence[CompliantExprT_contra]
+    ) -> tuple[CompliantLazyFrameAny, list[str], list[str]]:
```
Why are both of these using `CompliantLazyFrameAny`?

This isn't the lazy protocol, it should be like:

https://github.com/narwhals-dev/narwhals/blob/26d97e5184db6ddf242c69672250644d486fc663/narwhals/_compliant/group_by.py#L69

But I think it'd need to change to this `TypeVar`

https://github.com/narwhals-dev/narwhals/blob/26d97e5184db6ddf242c69672250644d486fc663/narwhals/_compliant/typing.py#L84

dangotbanned commented on narwhals/_compliant/group_by.py:
```diff
@@ -67,12 +71,69 @@ def compliant(self) -> CompliantFrameT_co:
         drop_null_keys: bool,
     ) -> None: ...
 
+    def _parse_keys(
+        self,
+        compliant_frame: CompliantLazyFrameAny,
+        keys: Sequence[CompliantExprT_contra] | Sequence[str],
+    ) -> tuple[CompliantLazyFrameAny, list[str], list[str]]:
+        if all(isinstance(k, str) for k in keys):
+            keys = cast("list[str]", list(keys))
+
+            return self._parse_string_keys(compliant_frame, keys=keys)
+        else:
+            keys = cast("list[CompliantExprT_contra]", list(keys))
+            return self._parse_expr_keys(compliant_frame, keys=keys)
```
Why are these being `cast`?

FBruzzesi replied:
I might be missing something but regardless of the `keys: Sequence[CompliantExprT_contra] | Sequence[str]` annotation and the `if all(isinstance(k, str) for k in keys):` check, mypy still misses the types.

Repro:

**t.py**

```py
from __future__ import annotations


def hello(some_list: list[int] | list[str]) -> str:
    if all(isinstance(v, str) for v in some_list):
        reveal_type(some_list)

        return "hello"
    else:
        reveal_type(some_list)
        return "world"

hello([1,2,3])
hello(["a", "b", "c"])
```

```terminal
mypy t.py
```

> t.py:6: note: Revealed type is "Union[builtins.list[builtins.int], builtins.list[builtins.str]]"
> 
> t.py:10: note: Revealed type is "Union[builtins.list[builtins.int], builtins.list[builtins.str]]"

dangotbanned replied:
Does this change anything?
```py
def hello(some_list: list[int] | list[str]) -> str:
    if is instance(some_list, list) and all(isinstance(v, str) for v in some_list):
```

Another difference between this and the actual code is the use of `Sequence[str]`.
I can see that tripping up `mypy`, since `str` is also `Sequence[str]` - so iterating over it doesn't narrow anything 😔

Buuuuuuut!

You can also just use a `TypeIs` function 😎
There should be something in `nw.utils` you could use or adapt?

FBruzzesi replied:
> Does this change anything?

Not really, even changing from Sequence to list

<details> <summary>screenshot</summary>
![image](https://github.com/user-attachments/assets/632fcd3c-2c8c-4e6f-ac9c-d3bf237ded48)
</details>

> You can also just use a TypeIs function 😎

Let me give it a shot 👀

dangotbanned replied:
https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051112276

@FBruzzesi lol that's my bad for making a suggestion without an IDE in https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051112276

`is_list_of` wasn't supposed to have the `type` part in the return 🤦‍♂️:

```py
# def is_list_of(obj: Any, tp: type[_T]) -> TypeIs[list[type[_T]]]:
def is_list_of(obj: Any, tp: type[_T]) -> TypeIs[list[_T]]:
```

Will fix this shortly

dangotbanned replied:
- [ ] #2397

FBruzzesi replied:
Thanks @dangotbanned, added a `is_sequence_of` for convenience and remove cast's calling.
Something is still missing but this is a first step

dangotbanned commented on narwhals/_polars/group_by.py:
```diff
@@ -19,45 +19,107 @@
 class PolarsGroupBy:
     _compliant_frame: PolarsDataFrame
-    _keys: Sequence[str]
+    _keys: list[PolarsExpr] | list[str]
```
This is incompatible with what you defined in:

https://github.com/narwhals-dev/narwhals/blob/69edc84be933dab7cdae9be9b139d81b9ef41172/narwhals/_compliant/group_by.py#L62-L64

FBruzzesi replied:
Even changing it to:

```diff
class CompliantGroupBy(Protocol38[CompliantFrameT_co, CompliantExprT_contra]): 
     _compliant_frame: Any 
-     _keys: list[str] 
+     _keys: list[str] | list[CompliantExprAny]
```

does not seem to suffice 🤔

dangotbanned replied:
@FBruzzesi there's a lot going on here 😅 

I'm going go through the [`mypy` yells](https://github.com/narwhals-dev/narwhals/actions/runs/14549212977/job/40818406408?pr=2325) and try to translate what the issues are and how to fix them 🙂

FBruzzesi replied:
Me right now 😂

![image](https://github.com/user-attachments/assets/a83b25bf-86c4-4d95-88c4-6c7438eb76a1)

dangotbanned replied:
> At this point I feel like I am witch hunting and unable to understand which protocol is not satisfied and (most importantly) the reason why 🥲

Okay forget about fixing `_keys` in isolation for now.

Let's try to understand how to satisfy the protocol (and which one?)

# CompliantDataFrame
`PolarsDataFrame` must satisfy this protocol.

The new part in this PR is `CompliantDataFrame.group_by`:

https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_compliant/dataframe.py#L138-L143

How does this compare to `PolarsDataFrame.group_by`?

https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_polars/dataframe.py#L380-L382

Looks good so far, but now we have to **prove** the following:
- `PolarsExpr` is [assignable](https://typing.python.org/en/latest/spec/glossary.html#term-assignable) to `CompliantExpr`
- `PolarsGroupBy` is assignable to `DataFrameGroupBy`

I'm going to skip `PolarsExpr` and assume it checks out for now.

# DataFrameGroupBy
This protocol is defined as an extension of [`CompliantGroupBy`](https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_compliant/group_by.py#L62), where we also require `__iter__` to be defined:
https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_compliant/group_by.py#L136-L140

To make this more digestible, let's just try checking everything **except** the types, with [`get_protocol_members`](https://docs.python.org/3/library/typing.html#typing.get_protocol_members):

```py
from typing_extensions import get_protocol_members

from narwhals._compliant.group_by import DataFrameGroupBy

>>> get_protocol_members(DataFrameGroupBy)
frozenset({'__iter__',
           '_compliant_frame',
           '_keys',
           '_output_key_names',
           '_parse_expr_keys',
           '_parse_keys',
           '_parse_string_keys',
           'agg',
           'compliant'})
```

## Is `PolarsGroupBy` assignable to `DataFrameGroupBy`?
https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_polars/group_by.py#L20

> [!WARNING]
> Of the members listed, we're **missing:**
> - `_output_key_names`
> - `_parse_expr_keys`
> - `_parse_keys`
> - `_parse_string_keys`

dangotbanned replied:
So we've identified a gap - now the questions are:

1. Does `PolarsGroupBy` work without these attributes/methods?
2. If yes, why are they part of `DataFrameGroupBy` if they are not required?

> [!TIP]
> By tightly defining protocols, we can avoid this issue

If we look at `DepthTrackingGroupBy`, it has lots of things defined that are not relevant to `polars` or lazy backends.

<details><summary>DepthTrackingGroupBy</summary>
<p>

https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_compliant/group_by.py#L143-L194

</p>
</details> 

If we defined these methods in `DataFrameGroupBy`, then `PolarsGroupBy` wouldn't be assignable to the protocol 🤯

FBruzzesi replied:
Thanks @dangotbanned that makes a lot of sense and I might start to get a sense of this!
The reason I added those methods in `CompliantGroupBy` is because that's the first/unique common parent of all the other derived classes (`DataFrameGroupBy`, `DepthTrackingGroupBy`, `EagerGroupBy` and `LazyGroupBy`).

What's the best path forward in your opinion? We could add a mixin class that defines such methods and let `EagerGroupBy` and `LazyGroupBy` inherit from that as well so that `CompliantGroupBy` and `DataFrameGroupBy` (which are returned in the compliant frames) remain untouched? Not sure how that interacts with protocols 🤔

MarcoGorelli replied:
🤞 hoping we can avoid mixins 🤞

dangotbanned replied:
> Thanks @dangotbanned that makes a lot of sense and I might start to get a grasp of this!

Music to my eyes! 😍 😍 😍 

> What's the best path forward in your opinion

We're closer to start solving issues - but not there just yet 🦫 

One more thing I want to explain in another thread

FBruzzesi replied:
> 🤞 hoping we can avoid mixins 🤞

& 

> Totally understandable 🙂

As in - another protocol is added and it solves this issue

dangotbanned commented on narwhals/_compliant/group_by.py:
```diff
@@ -67,12 +72,66 @@ def compliant(self) -> CompliantFrameT_co:
+        else:
+            return self._parse_expr_keys(compliant_frame, keys=keys)
+
+    @staticmethod
+    def _parse_expr_keys(
+        compliant_frame: CompliantFrameT, keys: Sequence[CompliantExprT_contra]
+    ) -> tuple[CompliantFrameT, list[str], list[str]]:
```
## Issue

Our problem is related to what happens *inside* the method

```py
narwhals/_compliant/group_by.py:122: error: Incompatible return value type (got
"tuple[CompliantDataFrame[Any, Any, Any] | CompliantLazyFrame[Any, Any], list[str], list[str]]", expected
"tuple[CompliantFrameT, list[str], list[str]]")  [return-value]
            return (
                   ^
```

If you split out the final `.with_columns` call, it reveals where the type becomes ambiguous

![image](https://github.com/user-attachments/assets/9fd146d3-4155-4f0a-8310-fca5a9897e61)

If you use *Go to definition*, we can see there are **two** possible signatures to match

![image](https://github.com/user-attachments/assets/46f7154d-2df8-43c3-882e-88f52a404611)

dangotbanned replied:
## The *oops my bad*
When I suggested (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051045009) - I meant to use that **everywhere** 🤦‍♂️ 

I'm gonna do that now, which gets `pyright` to point directly at the issue with `.with_columns`

dangotbanned replied:
Let me know if any of that made sense 😄

dangotbanned replied:
> [!IMPORTANT]
> I'm just going to work through identifying/explaining the problems **first**, before we think about how to solve them all

dangotbanned replied:
@FBruzzesi (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051483465) hold off on trying to fix it for now - we've got other issues (mainly https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051464670) that I wanna talk through first 🙂 

> but isn't that the definition of `CompliantFrameT` itself? 

No, the definition of `CompliantFrameT` is:
```py
CompliantFrameT = TypeVar("CompliantFrameT", bound=CompliantFrameAny)
```

The error message says we have:
```py
CompliantFrameAny: TypeAlias = "CompliantDataFrameAny | CompliantLazyFrameAny"
```

At the beginning of the method we **had** `CompliantFrameT`, but because there are two signatures (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051473155) calling `.with_columns` could return *either* one.

> Doing something like the following also does not help mypy:

Yeah that is just moving the same problem around - we'll get there don't worry 😎

dangotbanned replied:
@FBruzzesi @MarcoGorelli 

I've finished another essay in (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051497418) 😅

FBruzzesi replied:
This is really helpful! thanks for taking the time, I really really appreciate it!

Me ~5 hours ago: I won't invest much more time in this

Also me: write a script to compare objects vs protocols

FBruzzesi replied:
So... this is the remaining issue after adding all your other feedback in https://github.com/narwhals-dev/narwhals/pull/2325/commits/797cf742022b52b9c11bd6fa8bcf7fa25f14c6fb 🔥

FBruzzesi replied:
> No, the definition of `CompliantFrameT` is:
> 
> ```python
> CompliantFrameT = TypeVar("CompliantFrameT", bound=CompliantFrameAny)
> ```
> 
> The error message says we have:
> 
> ```python
> CompliantFrameAny: TypeAlias = "CompliantDataFrameAny | CompliantLazyFrameAny"
> ```
> 
> At the beginning of the method we **had** `CompliantFrameT`, but because there are two signatures ([#2325 (comment)](https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051473155)) calling `.with_columns` could return _either_ one.

I am still not fully sure that this error is clear. As the input is `CompliantFrameT` and `with_column` returns self from _both_ signatures, shouldn't the output expected to be the same?

dangotbanned replied:
I've been sitting on a solution for this 😏

Just wanted to make sure I could leak my brain elsewhere 😉

dangotbanned replied:
> [!NOTE]
> Hint: Have you looked at how `.drop_nulls` is avoiding the same issue?

dangotbanned replied:
Oh I found an even simpler way - forget about `drop_nulls`

dangotbanned replied:
Resolved in (https://github.com/narwhals-dev/narwhals/pull/2325/commits/b28da50c67bad1511042f6886660f15caca33781)

dangotbanned commented on narwhals/_compliant/group_by.py:
```diff
@@ -56,36 +59,90 @@
 
-class CompliantGroupBy(Protocol38[CompliantFrameT_co, CompliantExprT_contra]):
+class CompliantGroupBy(Protocol38[CompliantFrameT, CompliantExprT_contra]):
     _compliant_frame: Any
-    _keys: Sequence[str]
+    _keys: list[str] | list[CompliantExprAny]
```
# Why no fix??? 😞
@FBruzzesi 
 
So when I said this, I seem to have done a poor job explaining the problem (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051055721)

> This is incompatible with what you defined in:
> 
> https://github.com/narwhals-dev/narwhals/blob/69edc84be933dab7cdae9be9b139d81b9ef41172/narwhals/_compliant/group_by.py#L62-L64


Before, `PolarsGroupBy` was incompatible with `CompliantGroupBy`.


<table>
<tr>
<td align="center">
Before
</td>
<td align="center">
After
</td>
</tr>

<tr>
<td>


```py
class CompliantGroupBy:
    _compliant_frame: Any
    _keys: list[str]

class PolarsGroupBy:
    _compliant_frame: PolarsDataFrame
    _keys: list[PolarsExpr] | list[str]

class ArrowGroupBy:
    _compliant_frame: Any
    _keys: list[str]
```
  
</td>
<td>

```py
class CompliantGroupBy:
    _compliant_frame: Any
    _keys: list[str] | list[CompliantExprAny]

class PolarsGroupBy:
    _compliant_frame: PolarsDataFrame
    _keys: list[PolarsExpr] | list[str]

class ArrowGroupBy:
    _compliant_frame: Any
    _keys: list[str] | list[CompliantExprAny]
```
  
</td>
</tr>

</table>

But afterwards, everything else that depended on `_keys: list[str]` would be required to check that they don't have `list[CompliantExprAny]` to avoid these issues:

![image](https://github.com/user-attachments/assets/35c05051-7b14-421a-a123-5ec38e5ec0f8)

# Solution
I'm skipping straight to it this time 😱 

`CompliantGroupBy._keys` doesn't need a `Union`, since the `Polars*` cases are the only one's that *may* store expressions there.

> [!TIP]
> Instead of storing two types in `_keys`, just store an empty `list` for the expression branch

`_keys` is never used outside of `__init__`, so no need to lose sleep on it 😎 

https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_polars/group_by.py#L41-L59

dangotbanned replied:
# Solution (Even simpler)
You can remove the `_keys` attribute from the protocol.

The only reference to it (in a protocol) is here in `LazyGroupBy`

https://github.com/narwhals-dev/narwhals/blob/caaa0924007be740a9aef1f41c4142e99dc68ff5/narwhals/_compliant/group_by.py#L218-L219

So you can just define this and leave the other bits to do with `_keys` what they want 🙂:

```py
class LazyGroupBy(
    CompliantGroupBy[CompliantLazyFrameT, LazyExprT_contra],
    Protocol38[CompliantLazyFrameT, LazyExprT_contra, NativeExprT_co],
):
    _keys: list[str]
```

> [!IMPORTANT]
> Keep the `keys` argument in the constructors though - that one is needed outside

dangotbanned commented on narwhals/_compliant/group_by.py:
```diff
@@ -30,6 +32,8 @@
 if TYPE_CHECKING:
     from typing_extensions import TypeAlias
 
+    _SameFrameT = TypeVar("_SameFrameT", CompliantDataFrameAny, CompliantLazyFrameAny)
```
https://typing.python.org/en/latest/spec/generics.html#introduction

Very similar thing going on in `nw.concat` 
https://github.com/narwhals-dev/narwhals/blob/db1c5a34784aca2cd1951da8c664f17a8f47cad3/narwhals/functions.py#L72-L75

dangotbanned replied:
@FBruzzesi FYI this wasn't what I had planned, but seemed simpler 😄 (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051535719)

My other idea was moving the `with_columns` into `__init__`, like we already have for `drop_nulls`

https://github.com/narwhals-dev/narwhals/blob/b28da50c67bad1511042f6886660f15caca33781/narwhals/_arrow/group_by.py#L51

That would've shown the same issue if it were used in the protocol that had a `Union`.

We don't see an issue in the concrete class, because the `TypeVar` has been resolved by an actual class

https://github.com/narwhals-dev/narwhals/blob/b28da50c67bad1511042f6886660f15caca33781/narwhals/_arrow/group_by.py#L43

Hope that makes sense :sweat_smile:

dangotbanned commented on narwhals/utils.py:
```diff
@@ -1302,7 +1302,11 @@ def is_list_of(obj: Any, tp: type[_T]) -> TypeIs[list[_T]]:
 def is_sequence_of(obj: Any, tp: type[_T]) -> TypeIs[Sequence[_T]]:
     # Check if an object is a sequence of `tp`, only sniffing the first element.
-    return bool(isinstance(obj, Sequence) and obj and isinstance(obj[0], tp))
+    return bool(
+        is_sequence_but_not_str(obj)
+        and (first := next(iter(obj), None))
+        and isinstance(first, tp)
+    )
```
Forgot to mention (https://github.com/narwhals-dev/narwhals/pull/2325/commits/66a24f2b1887d5eb9e447c0d096da48334cafc8a) this *also* fixes `str` passing through

dangotbanned commented on narwhals/_spark_like/dataframe.py:
```diff
@@ -187,7 +187,7 @@ def _iter_columns(self) -> Iterator[Column]:
     @property
     def columns(self) -> list[str]:
-        return list(self.schema)
+        return list(self.schema) if self._cached_schema else list(self.native.columns)
```
Does the second one need the `list` call?

I didn't see a mention of it in https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2050992665

https://github.com/eakmanrq/sqlframe/blob/ce3db1f1500e3fd28494860dce1d421a05b8e26e/sqlframe/base/dataframe.py#L294

dangotbanned commented on narwhals/_spark_like/group_by.py:
```diff
@@ -16,19 +16,23 @@ class SparkLikeLazyGroupBy(LazyGroupBy["SparkLikeLazyFrame", "SparkLikeExpr", "C
         drop_null_keys: bool,
     ) -> None:
-        self._compliant_frame = df.drop_nulls(subset=None) if drop_null_keys else df
-        self._keys = list(keys)
+        frame, self._keys, self._output_key_names = self._parse_keys(df, keys=keys)
+        self._compliant_frame = (
+            frame.drop_nulls(subset=self._keys) if drop_null_keys else frame
+        )
```
```suggestion
        self._compliant_frame = frame.drop_nulls(self._keys) if drop_null_keys else frame
```

dangotbanned commented on narwhals/_spark_like/group_by.py:
```diff
@@ -16,19 +16,23 @@ class SparkLikeLazyGroupBy(LazyGroupBy["SparkLikeLazyFrame", "SparkLikeExpr", "C
-        self._compliant_frame = df.drop_nulls(subset=None) if drop_null_keys else df
-        self._keys = list(keys)
+        frame, self._keys, self._output_key_names = self._parse_keys(df, keys=keys)
+        self._compliant_frame = (
+            frame.drop_nulls(subset=self._keys) if drop_null_keys else frame
+        )
```
```suggestion
        self._compliant_frame = frame.drop_nulls(self._keys) if drop_null_keys else frame
```

FBruzzesi replied:
ruff doesn't like this (I think it would it around 95 chars line length)

dangotbanned replied:
Huh weird - mine was autocollapsing, all I wanted to do was remove the keyword - no biggie

FBruzzesi replied:
Oh nevermind, my bad!

dangotbanned commented on narwhals/_expression_parsing.py:
```diff
@@ -183,9 +181,18 @@ class ExpansionKind(Enum):
     MULTI_UNNAMED = auto()
     """e.g. `nw.all()`, nw.nth(0, 1)"""
 
+    def is_single(self) -> bool:  # pragma: no cover
+        return self is ExpansionKind.SINGLE
+
+    def is_multi_named(self) -> bool:  # pragma: no cover
+        return self is ExpansionKind.MULTI_NAMED
+
```
No need to add these yet if they aren't used
```suggestion
```

dangotbanned commented on tests/group_by_test.py:
```diff
@@ -455,3 +463,144 @@ def test_fancy_functions(constructor: Constructor) -> None:
+        # See: https://github.com/pola-rs/polars/issues/22238
+        request.applymarker(pytest.mark.xfail)
+
+    if (
+        "polars" in str(constructor)
+        and drop_null_keys
+        and any(key._metadata.expansion_kind.is_multi_output() for key in keys)
+    ):
+        request.applymarker(pytest.mark.xfail)
```
@FBruzzesi what is the reason behind this one failing?

Also, AFAICT you've added a method that is only used as part of this `xfail` 🤔 

https://github.com/narwhals-dev/narwhals/blob/f0792441319611bcef27f9d14c32465f9aa52fce/narwhals/_expression_parsing.py#L193

I think this point is similar (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2040736302)

dangotbanned commented on narwhals/dataframe.py:
```diff
@@ -1536,18 +1537,27 @@ def group_by(
+        if all(isinstance(key, str) for key in flat_keys):
+            return GroupBy(self, flat_keys, drop_null_keys=drop_null_keys)
+
+        from narwhals import col
+        from narwhals.expr import Expr
+        from narwhals.series import Series
+
+        _keys = [k if isinstance(k, (Expr, Series)) else col(k) for k in flat_keys]
+        expr_flat_keys, kinds = self._flatten_and_extract(*_keys)
```
I've made this change locally and it only caused a single test to fail.

```suggestion
        expr_flat_keys, kinds = self._flatten_and_extract(*flat_keys)
```

I think the error message is reasonable and this is in `main` so I thought the plan was to raise for this anyway?

Did I miss/forget something? :dizzy_face:

```
E       narwhals.exceptions.InvalidIntoExprError: Expected an object which can be converted into an expression, got <class 'bool'>
E       
E       Hint:
E       - if you were trying to select a column which does not have a string
E         column name, then you should explicitly use `nw.col`.
E         For example, `df.select(nw.col(0))` if you have a column named `0`.
E       - if you were trying to create a new literal column, then you 
E         should explicitly use `nw.lit`.
E         For example, `df.select(nw.lit(0))` if you want to create a new
E         column with literal value `0`.

narwhals\dataframe.py:437: InvalidIntoExprError
```

FBruzzesi replied:
> I've made this change locally and it only caused a single test to fail.

This was to handle pandas column names that can be bool and int

MarcoGorelli replied:
If were starting from scratch I think we could force users to wrap non string column names in nw.col , but seeing as we've allowed non string column names as arguments to group_by so far, we need to keep allowing that

dangotbanned replied:
No worries folks

dangotbanned commented:
This seems to be flaky https://github.com/narwhals-dev/narwhals/actions/runs/14552778275/job/40825960555?pr=2325

It was failing earlier today as well https://github.com/narwhals-dev/narwhals/actions/runs/14551486696/job/40823184992?pr=2325

FBruzzesi commented:
> This seems to be flaky https://github.com/narwhals-dev/narwhals/actions/runs/14552778275/job/40825960555?pr=2325
> 
> > FAILED tests/group_by_test.py::test_key_with_nulls_iter[pandas]
> 
> It was failing earlier today as well https://github.com/narwhals-dev/narwhals/actions/runs/14551486696/job/40823184992?pr=2325

Yes! I think it is a quite specific pandas version! I can give it a try tomorrow and investigate deeper

dangotbanned commented on narwhals/dataframe.py:
```diff
@@ -1488,12 +1488,13 @@ def filter(
         Arguments:
-            *keys: Column(s) to group by. Accepts multiple columns names as a list.
+            *keys: Column(s) to group by. Accepts expression input. Strings are parsed as column names.
+                Multiple columns should can be passed as a list.
```
```suggestion
                Multiple columns can be passed as a list.
```

dangotbanned commented on narwhals/dataframe.py:
```diff
@@ -2770,14 +2780,14 @@ def filter(
-                Column(s) to group by. Accepts expression input. Strings are
-                parsed as column names.
+                Column(s) to group by. Accepts expression input. Strings are parsed as column names.
+                Multiple columns should can be passed as a list.
```
```suggestion
                Multiple columns can be passed as a list.
```

MarcoGorelli commented on narwhals/_arrow/group_by.py:
```diff
@@ -141,10 +143,21 @@ def __iter__(self) -> Iterator[tuple[Any, ArrowDataFrame]]:
             t = self.compliant._with_native(
                 table.filter(pc.equal(table[col_token], v)).drop([col_token])
             )
+            # Should we add a `.head(1)` before calling `simple_select` and `row`?
```
probably not, why?

MarcoGorelli commented on tests/group_by_test.py:
```diff
@@ -455,3 +461,144 @@ def test_fancy_functions(constructor: Constructor) -> None:
+    df = nw.from_native(constructor(data))
+
+    context: Any
+    if isinstance(df, nw.LazyFrame) and any(
+        expr._metadata.kind.is_filtration() for expr in keys
+    ):
+        context = pytest.raises(
+            LengthChangingExprError,
+            match="Length-changing expressions are not supported for use in LazyFrame",
+        )
+    elif isinstance(df, nw.LazyFrame) and any(
+        expr._metadata.kind.is_window() for expr in keys
+    ):
+        context = pytest.raises(
+            OrderDependentExprError,
+            match="Order-dependent expressions are not supported for use in LazyFrame",
+        )
+    else:
+        context = pytest.raises(
+            ComputeError,
+            match="Group by is not supported with keys that are not transformation expressions",
+        )
```
not sure about this, can we have the exception be on of the arguments we parametrise over? i think we can skip `match=`

dangotbanned replied:
Adding to this, I think this is another case of (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2040736302)

We'd need to rewrite this test if we change the internal structure of `ExprMetadata` currently

FBruzzesi replied:
This is definitly a weird territory and the reason for it is that for lazyframes we control early for order dependent expressions. For eager ones we end up in the check performed in the group_by method and exceptions are consistent/same in each case.

> can we have the exception be on of the arguments we parametrise over?

Yes and no! Meaning that it also depends on the constructor (lazy vs eager) 🤔 I will push an idea and get back asking for feedback

FBruzzesi replied:
Check changes in https://github.com/narwhals-dev/narwhals/pull/2325/commits/62f5749b0ed9fa5684f5f2100ba2b1f0c0e85266

dangotbanned replied:
Resolved in (https://github.com/narwhals-dev/narwhals/commit/62f5749b0ed9fa5684f5f2100ba2b1f0c0e85266)

MarcoGorelli commented on narwhals/_polars/group_by.py:
```diff
@@ -18,45 +19,93 @@
     _compliant_frame: PolarsLazyFrame
-    _keys: Sequence[str]
+    _grouped: NativeLazyGroupBy
+    _drop_null_keys: bool
+    _output_names: Sequence[str]
 
     @property
     def compliant(self) -> PolarsLazyFrame:
         return self._compliant_frame
 
     def __init__(
-        self, df: PolarsLazyFrame, keys: Sequence[str], /, *, drop_null_keys: bool
+        self,
+        df: PolarsLazyFrame,
+        keys: Sequence[PolarsExpr] | Sequence[str],
+        /,
+        *,
+        drop_null_keys: bool,
     ) -> None:
         self._compliant_frame = df
-        self._keys = list(keys)
-        df = df.drop_nulls(keys) if drop_null_keys else df
-        self._grouped: NativeLazyGroupBy = df._native_frame.group_by(keys)
+        self._drop_null_keys = drop_null_keys
+        self._output_names = []
+
+        if is_sequence_of(keys, str):
+            self._output_names = keys
+            self._grouped = self.compliant.native.group_by(keys)
+        else:
+            if drop_null_keys:
+                self._output_names = flatten(
+                    [
+                        arg.native.meta.root_names()
+                        if arg.native.meta.has_multiple_outputs()
+                        else arg.native.meta.output_name()
+                        for arg in keys
+                    ]
+                )
+            self._grouped = self.compliant.native.group_by(arg.native for arg in keys)
```
using `Expr.meta` kinda scares  me here

do we need to intercept Polars' behaviour here, or can we just leave it to do its thing?

FBruzzesi replied:
I am trying to figure out the output key names when keys are expressions so that we can drop null keys if such flag is true. I couldn't find a better way of achieving that from the docs 🤔

Isn't `Expr.meta` stable enough?

MarcoGorelli replied:
not sure, i've encountered bugs in it unfortunately

i'll investigate - if it's unfeasible we may want to consider only allowing expressions when `drop_null_keys` is `False` (the default)?

dangotbanned replied:
> Isn't `Expr.meta` stable enough?

I think so - but maybe @MarcoGorelli has insider knowledge 😉 

> I am trying to figure out the output key names when keys are expressions so that we can drop null keys if such flag is true.

Do we need to support `drop_null_keys` when we're provided `Expr` input?
I was thinking about that while reviewing, since it seems like that's where a lot of the complexity in the PR is at.

IIUC, that parameter was added as a `pandas` optimization for `plotly` (#1256) - but I can't see it performant when we need to do all the extra work with expressions 🤔 

[`pl.DataFrame.drop_nulls`](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.drop_nulls.html) doesn't accept `Expr` - so I think that is a *possible* out for us 🙏

dangotbanned replied:
> i'll investigate - if it's unfeasible we may want to consider only allowing expressions when `drop_null_keys` is `False` (the default)?

@MarcoGorelli 🌊📏 
Promise I didn't see yours while writing (https://github.com/narwhals-dev/narwhals/pull/2325#discussion_r2051691688) 😄

FBruzzesi replied:
Thanks both!

@MarcoGorelli 

> if it's unfeasible we may want to consider only allowing expressions when drop_null_keys is False (the default)?

Just to double check: do you mean for polars only on generally for all backends?

@dangotbanned 

> IIUC, that parameter was added as a pandas optimization for plotly (https://github.com/narwhals-dev/narwhals/issues/1256) - but I can't see it performant when we need to do all the extra work with expressions 🤔

Not necessarily optimizing for pandas, rather, that's the default behavior for [pandas groupby](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html) and the expectation for plotly users.

dangotbanned replied:
Thanks @FBruzzesi for clarifying the `pandas` part.

> Just to double check: do you mean for polars only on generally for all backends?

I know this was directed at @MarcoGorelli, but my view would've been all backends.
The tricky part for `pandas` might be that we'd need an exception for the non-string columns case

FBruzzesi replied:
It might get simpler that expected actually. I will push a commit to show

FBruzzesi replied:
See https://github.com/narwhals-dev/narwhals/pull/2325/commits/a16ec2a7d2e8bb72b7da2586f2e65c0e6d1baf37

The idea is: instead of checking for specific types, I just check if any expr (or series in the eager case) are passed. That combined with the fact of having `drop_null_keys=True` will raise a `NotImplementedError`

dangotbanned replied:
@FBruzzesi great!

I've added this - but the tests will still need adapting to the new typing

- [feat(typing): Add overloads for `drop_null_keys` string only](https://github.com/narwhals-dev/narwhals/pull/2325/commits/bb22b1b178488b967639ac04ae662b1bc66007c4)

dangotbanned commented on narwhals/group_by.py:
```diff
@@ -14,18 +15,26 @@
+    def __init__(
+        self,
+        df: DataFrameT,
+        keys: Sequence[str] | Sequence[CompliantExprAny],
+        *,
+        drop_null_keys: bool,
```
```suggestion
        keys: Sequence[str] | Sequence[CompliantExprAny],
        /,
        *,
        drop_null_keys: bool,
```

dangotbanned replied:
Both `df` and `keys` were *positional-only* before - so this keeps it that way 🙂 

I remember @MarcoGorelli mentioning wanting to rename `keys` to `by` to [match upstream](https://docs.pola.rs/api/python/stable/reference/dataframe/api/polars.DataFrame.group_by.html).
Keeping things positional will make that transition seamless 😎

dangotbanned commented on narwhals/group_by.py:
```diff
@@ -98,11 +107,17 @@ def __iter__(self) -> Iterator[tuple[Any, DataFrameT]]:
+    def __init__(
+        self,
+        df: LazyFrameT,
+        keys: Sequence[str] | Sequence[CompliantExprAny],
+        *,
+        drop_null_keys: bool,
```
```suggestion
        keys: Sequence[str] | Sequence[CompliantExprAny],
        /,
        *,
        drop_null_keys: bool,
```

dangotbanned commented on tests/group_by_test.py:
```diff
@@ -455,3 +459,117 @@ def test_fancy_functions(constructor: Constructor) -> None:
+    ],
+    ids=range(5),
+)
+@pytest.mark.parametrize("drop_null_keys", [True, False])
+def test_group_by_expr(
+    request: pytest.FixtureRequest,
+    constructor: Constructor,
+    keys: list[nw.Expr],
+    aggs: list[nw.Expr],
+    expected: dict[str, list[Any]],
+    sort_by: list[str],
+    *,
+    drop_null_keys: bool,
+) -> None:
+    request_id = request.node.callspec.id
+    if (
+        POLARS_COLLECT_STREAMING_ENGINE
+        and request_id.startswith("polars[lazy]")
+        and request_id.endswith("0")
+    ):
+        # Blocked by upstream issue as of polars==1.27.1
+        # See: https://github.com/pola-rs/polars/issues/22238
+        request.applymarker(pytest.mark.xfail)
+
+    context = (
+        pytest.raises(
+            NotImplementedError,
+            match="drop_null_keys cannot be True when keys contains Expr",
+        )
+        if drop_null_keys
+        else nullcontext()
```
Following
- (https://github.com/narwhals-dev/narwhals/pull/2325/commits/a16ec2a7d2e8bb72b7da2586f2e65c0e6d1baf37)
- (https://github.com/narwhals-dev/narwhals/pull/2325/commits/bb22b1b178488b967639ac04ae662b1bc66007c4)

I think we can just move the `drop_null_keys=True` to another test

There's no need anymore to check all the same parameterized tests, we could just do something like the `IntoExpr | Iterable[IntoExpr]` variants (minus `str`)

dangotbanned replied:
Would also simplify fixing this https://github.com/narwhals-dev/narwhals/actions/runs/14559734698/job/40841407746?pr=2325

dangotbanned replied:
Thanks @FBruzzesi 

Resolved in (https://github.com/narwhals-dev/narwhals/pull/2325/commits/db9175fe9ac997fbde3013b3d0d09a0dde75a65f)

dangotbanned commented on tests/group_by_test.py:
```diff
@@ -455,3 +459,121 @@ def test_fancy_functions(constructor: Constructor) -> None:
+            {"a": [1, 1, 2], "y": [0.5, 1.5, 1], "x": [1, 4, 5]},
+            ["a", "y"],
+        ),
+        (
+            [nw.col("a").abs().alias("y")],
+            [nw.all().sum().name.suffix("c")],
+            {"y": [1, 2], "ac": [1, 4], "xc": [5, 5]},
+            ["y"],
+        ),
```
Would there be any benefit to having another test using `narwhals.selectors`?
We seem to use `nw.all()` in an aggregation - which is close.

I'd be interested to see one using *purely* selectors in both the `group_by(...)` and `.agg(...)` clauses (if that works 😏)

dangotbanned replied:
@FBruzzesi fantastic!

dangotbanned commented on narwhals/_pandas_like/dataframe.py:
```diff
@@ -655,7 +655,12 @@ def collect(
 
     # --- actions ---
-    def group_by(self, *keys: str, drop_null_keys: bool) -> PandasLikeGroupBy:
+    def group_by(
+        self,
+        keys: Sequence[str] | Sequence[PandasLikeExpr],
+        *,
+        drop_null_keys: bool,
+    ) -> PandasLikeGroupBy:
```
```suggestion
    def group_by(
        self, keys: Sequence[str] | Sequence[PandasLikeExpr], *, drop_null_keys: bool
    ) -> PandasLikeGroupBy:
```

dangotbanned commented on narwhals/_duckdb/expr.py:
```diff
@@ -345,13 +345,7 @@ def alias_output_names(names: Sequence[str]) -> Sequence[str]:
-            backend_version=self._backend_version,
-            version=self._version,
-        )
+        return self._with_alias_output_names(alias_output_names)
 
```
Very nice 😍

dangotbanned commented on narwhals/_compliant/group_by.py:
```diff
@@ -83,9 +90,61 @@ class DataFrameGroupBy(
+    ) -> tuple[_SameFrameT, list[str], list[str]]:
+        """Parses key expressions to set up `.agg` operation with correct information.
+
+        Since keys are expressions, it's possible to incour in aliases that match
```
> incour

Is this a typo?

FBruzzesi replied:
Should have been "incur", but I just rephrased it

dangotbanned commented on narwhals/_arrow/group_by.py:
```diff
@@ -141,10 +143,20 @@ def __iter__(self) -> Iterator[tuple[Any, ArrowDataFrame]]:
+
+        original_table = self._df
+        if self._drop_null_keys:
+            original_table = original_table.drop_nulls(subset=list(mapping.values()))
```
```suggestion
            original_table = original_table.drop_nulls(list(mapping.values()))
```

FBruzzesi commented on narwhals/dataframe.py:
```diff
@@ -1487,13 +1487,24 @@ def filter(
-            *keys: Column(s) to group by. Accepts multiple columns names as a list.
+            *keys: Column(s) to group by. Accepts expression input. Strings are parsed as column names.
+                Multiple columns can be passed as a list.
             drop_null_keys: if True, then groups where any key is null won't be included
                 in the result.
 
```
Added one extra example in both eager and lazy

FBruzzesi commented:
> Great work @FBruzzesi 🎉
> 
> I've left some comments/suggestions, but nothing blocking.

Thanks @dangotbanned - I am a bit short on time this week and the next - so apologies for my slow feedback loop.
I should have addressed all your comments/suggestion 😎

> This is an exciting step towards _expressifying_ more of `narwhals` 🚀

It definitly is - exciting times!

dangotbanned reviewed (approved):
https://github.com/narwhals-dev/narwhals/pull/2325#issuecomment-2826617619

Thanks again @FBruzzesi 

I'll leave open until the end of the day in case @MarcoGorelli wants time chime in - but otherwise I'll merge later ❤️

MarcoGorelli commented:
thanks both, really amazing stuff!

i'll just trying breaking this a bit more, then i'd say we can ship it

MarcoGorelli commented on narwhals/_compliant/group_by.py:
```diff
@@ -83,9 +90,61 @@ class DataFrameGroupBy(
+            key.name.suffix(suffix_token)
+            if key._metadata is not None
+            and key._metadata.expansion_kind is ExpansionKind.MULTI_NAMED
+            # otherwise it's single named and we can use Expr.alias
+            else key.alias(f"{new_name}{suffix_token}")
+            for key, new_name in zip(keys, output_names)
```
not necessarily, no

example:
```python
import polars as pl
import duckdb
import pyarrow as pa
import pandas as pd
import narwhals as nw

data = {'a': [1,2,3], 'b': [4,5,6], 'c': [7., 8., 9.]}

df = pl.DataFrame(data)

def func(df):
    return print(nw.from_native(df).lazy().group_by(
        nw.selectors.by_dtype(nw.Int64)
    ).agg(
        nw.col('c').mean().name.suffix('_agg')
    ).sort('a').collect('polars'))

print(func(df))
print(func(df.to_pandas()))
print(func(df.to_arrow()))
print(func(duckdb.table('df')))
```

MarcoGorelli commented on narwhals/_arrow/group_by.py:
```diff
@@ -141,10 +144,20 @@ def __iter__(self) -> Iterator[tuple[Any, ArrowDataFrame]]:
+
         table = table.add_column(i=0, field_=col_token, column=key_values)
+
+        original_table = self._df
+        if self._drop_null_keys:
+            original_table = original_table.drop_nulls(list(mapping.values()))
```
we're now doing two `drop_nulls`, whereas previously we only did it once

MarcoGorelli replied:
this isn't used anywhere, it can be removed

dangotbanned replied:
Ah yeah nice one @MarcoGorelli 

I forgot to follow-up if anything else wasnt needed sincehttps://github.com/narwhals-dev/narwhals/pull/2325/commits/a16ec2a7d2e8bb72b7da2586f2e65c0e6d1baf37

MarcoGorelli reviewed (approved):
I reverted a change to an existing test - in general i'd suggest to just add a new test if you're testing something different

For anything this close to the library core and to expression parsing, please do wait for me to take a final look

Anyway, this looks great, thanks @FBruzzesi and @dangotbanned ! You have a lot to be proud of here. I'm certainly proud that all together we've arrived at such a simple-looking solution

type checking does pass in CI, but there is the occasional squiggly line in vscode. happy to resolve that separately though

![](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExbWwzdDJkaGJvdG1ncXZyb2dlNXhhaHdyZWYyYW5wendoYjJ1eGc4dyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/xE2xnByqrmBzP9eyuE/giphy.gif)

dangotbanned commented:
@MarcoGorelli would we be able to set up something to preserve co-authors? 

I think I'm missing again 😔
https://github.com/narwhals-dev/narwhals/commit/89448231342acda3358efea20419203e6ea35c56

MarcoGorelli commented:
ah so sorry 😩 😳 

the combination of 124 commit messages was too long so i removed the whole thing, but it looks like i should've kept the "co-authored by" part at the end for co-authors to show up

I'll be very careful to always do that in the future 👍
